components/feature/qr/src/main/java/mozilla/components/feature/qr/views/CustomViewFinder.kt

```diff
@@ -0,0 +1,284 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.qr.views
+
+import android.content.Context
+import android.content.res.Configuration
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.Path
+import android.graphics.Rect
+import android.os.Build
+import android.text.Layout
+import android.text.StaticLayout
+import android.text.TextPaint
+import android.util.AttributeSet
+import android.util.DisplayMetrics
+import android.util.TypedValue
+import android.view.View
+import android.view.ViewGroup
+import androidx.annotation.ColorInt
+import androidx.annotation.Px
+import androidx.appcompat.widget.AppCompatImageView
+import androidx.annotation.StringRes
+import androidx.core.content.ContextCompat
+import androidx.core.text.HtmlCompat
+import kotlin.math.min
+import kotlin.math.roundToInt
+
+/**
+ * A [View] that shows a ViewFinder positioned in center of the camera view and draws an Overlay
+ */
+@Suppress("LargeClass", "TooManyFunctions")
+class CustomViewFinder @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null
+) : AppCompatImageView(context, attrs) {
+    private var messageResource: Int? = null
+    private val overlayPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
+    private val viewFinderPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
+    private var viewFinderPath: Path = Path()
+    private var viewFinderPathSaved: Boolean = false
+    private var overlayPath: Path = Path()
+    private var overlayPathSaved: Boolean = false
+
+    private lateinit var viewFinderRectangle: Rect
+    private var viewFinderCornersSize: Float = 0f
+    private var viewFinderCornersRadius: Float = 0f
+    private var viewFinderTop: Float = 0f
+    private var viewFinderLeft: Float = 0f
+    private var viewFinderRight: Float = 0f
+    private var viewFinderBottom: Float = 0f
+    private var normalizedRadius: Float = 0f
+
+    private var scanMessageLayout: StaticLayout? = null
+    private lateinit var messageTextPaint: TextPaint
+
+    init {
+        isSaveEnabled = true
+        overlayPaint.style = Paint.Style.FILL
+        viewFinderPaint.style = Paint.Style.STROKE
+        overlayPath.fillType = Path.FillType.EVEN_ODD
+        viewFinderPath.fillType = Path.FillType.EVEN_ODD
+
+        this.layoutParams = ViewGroup.LayoutParams(
+            ViewGroup.LayoutParams.MATCH_PARENT,
+            ViewGroup.LayoutParams.MATCH_PARENT)
+        this.setOverlayColor(DEFAULT_OVERLAY_COLOR)
+        this.setViewFinderColor(DEFAULT_VIEWFINDER_COLOR)
+        this.setViewFinderStroke(DEFAULT_VIEWFINDER_THICKNESS_DP.dpToPx(resources.displayMetrics))
+        this.setViewFinderCornerRadius(DEFAULT_VIEWFINDER_CORNERS_RADIUS_DP.dpToPx(resources.displayMetrics))
+    }
+
+    private fun computeViewFinderRect(width: Int, height: Int) {
```

Comment:
Tricky to unit test view code, but it seems reasonable to add a test for this method and maybe also `drawOverlay/drawViewFinder` if `Path` can be spied on?

Reply:
I will give it a try. :)

Reply:
I added some unit tests. I did not find a way to spy on the path or canvas. Also, I avoided using Robolectric. There are some framework (https://facebook.github.io/screenshot-tests-for-android/) but I don't think it's worth adopting it just for these tests.

Reply:
ok, thanks!